### ISSUE TITLE
HMR does not work for JS files 

### ISSUE THREAD

**Issue by jprosevear** [vitejs/vite-plugin-react#191]
### Describe the bug

Using the jsx loader for js as files as per:
https://github.com/vitejs/vite/discussions/3448#discussioncomment-749919

HMR does not work.  Full page reloads happen instead.

Originally thought this might be solved by https://github.com/vitejs/vite-plugin-react/pull/188#issuecomment-1615988252 but discussion there showed it was not.

This might be marked won't fix because Vite recommends the jsx/tsx extensions.  However I thought it was worth a specific bug ticket even if its closed so there is a record of the decision.

Please note that create-react-app (where I'm porting from), nextjs and gatsby all support .js and this would help minimize the changes to port apps.

### Reproduction

https://stackblitz.com/edit/vitejs-vite-p5o1nn?file=src%2FApp.js

### Steps to reproduce

<img width="408" alt="image" src="https://github.com/vitejs/vite-plugin-react/assets/699616/247d1073-8f23-4279-8a77-386b2f6dd6da">
(Changing something in App.js)

### System Info

```shell
See stack blitz
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite-plugin-react/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite-plugin-react/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite-plugin-react/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite-plugin-react/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ArnaudBarre** [comment]
This has been fixed in https://github.com/vitejs/vite-plugin-react/pull/188 I think.
Anyway this will not be better supported in Vite. esbuild, swc and bun agreed to not enable the JSX parser on JS files. Plus most codebases are now in TS and the extension is mandatory.  Migrating now will simplify a lot the setup of future tools. There are plenty of one line command available on the gitst/stack overflow to migrate.

**Comment by jprosevear** [comment]
> This has been fixed in #188 I think. Anyway this will not be better supported in Vite. esbuild, swc and bun agreed to not enable the JSX parser on JS files. Plus most codebases are now in TS and the extension is mandatory. Migrating now will simplify a lot the setup of future tools. There are plenty of one line command available on the gitst/stack overflow to migrate.

It does not seem to have been fixed as per:
https://github.com/vitejs/vite-plugin-react/pull/188#issuecomment-1615988252
https://github.com/vitejs/vite-plugin-react/pull/188#issuecomment-1616033891

**Comment by ArnaudBarre** [comment]
Yeah you will need to use an explicit  `import React`

**Comment by jprosevear** [comment]
@ArnaudBarre @csr632 https://github.com/vitejs/vite/discussions/3448#discussioncomment-5681381 plus the unreleased version of this plugin solves the issue.

The fix in #188 is required so that the "react/jsx-dev-runtime" is used to set useFastRefresh correctly

```
 (isJSX ||
          (opts.jsxRuntime === 'classic'
            ? importReactRE.test(code)
            : code.includes(devRuntime)))
```

(code.includes in the correct order in the ternerary)
